### ISSUE TITLE
[BUG] Refine: uncaught `IndexError: index 0 is out of bounds for axis 0 with size 0` when adding `--divergence-units mutations`

### ISSUE THREAD

**Issue by mdperry** [nextstrain/augur#1013]
### Current Behavior
I was able to build clocked trees (from UShER) using four augur commands, one after the other.  I wanted the X-axis of the plotted trees to use the number of mutations (instead of percent divergence, or whatever the default is), so I added this flag to my augur refine command: '--divergence-units mutations'.  Now it dies with a Python stack trace.  The final line is above.  I have a batch of 10 trees, the first three seem to proceed without complaint but I think it is dying while processing the fourth input.  I base this on the regular augur stderr.

### Expected behavior
Adding an extra flag would not cause the program to die.

### How to reproduce
Steps to reproduce the current behavior:
1. This is the command I am running as a shell command inside Terra:
`augur refine -t newfile.nwk -a newfile.vcf --vcf-reference ~{reference} --metadata user_subset_metadata.tsv --timetree --root 'min_dev' --coalescent 'opt' --output-tree augur.tree.nwk --output-node augur.branch_lengths.json --divergence-units 'mutations'`

The ~{reference} is WDL's version of variable interpretation, and it gets expanded to 'wuhCov1.fa' a fasta file with one of the SARS-CoV-2 'wild-type' RefSeqs.  user_subset_metadata.tsv is a table I created from an UShER subtree.json file that can be displayed on auspice.us.  I harvested all of the key => value pairs from there using a custom Perl script.  newfile.nwk and newfile.vcf are both created by matUtils extract (in the UShER suite of tools), after feeding in a very large protobuf tree file (also built by UShER), and the same metadata file, and a list of samples, oh and a translation table from the tree as well.

2. This is the actual error message text that I copied and pasted from the stderr in my Terra execution directory for this task:
/usr/local/lib/python3.7/site-packages/treetime/gtr.py:955: RuntimeWarning: overflow encountered in exp
  return np.exp(self.mu * t * self.eigenvals)
/usr/local/lib/python3.7/site-packages/scipy/optimize/optimize.py:2621: RuntimeWarning: invalid value encountered in double_scalars
  w = xb - ((xb - xc) * tmp2 - (xb - xa) * tmp1) / denom
/usr/local/lib/python3.7/site-packages/scipy/optimize/optimize.py:2214: RuntimeWarning: invalid value encountered in double_scalars
  tmp1 = (x - w) * (fx - fv)
/usr/local/lib/python3.7/site-packages/scipy/optimize/optimize.py:2215: RuntimeWarning: invalid value encountered in double_scalars
  tmp2 = (x - v) * (fx - fw)
/usr/local/lib/python3.7/site-packages/treetime/treeanc.py:938: RuntimeWarning: invalid value encountered in subtract
  normalized_profile = (self.tree.root.joint_Lx.T - self.tree.root.joint_Lx.max(axis=1)).T
/usr/local/lib/python3.7/site-packages/treetime/distribution.py:150: RuntimeWarning: invalid value encountered in subtract
  yvals -= self._peak_val
Traceback (most recent call last):
  File "/nextstrain/augur/augur/__init__.py", line 81, in run
    return args.__command__.run(args)
  File "/nextstrain/augur/augur/refine.py", line 219, in run
    covariance=args.covariance, resolve_polytomies=(not args.keep_polytomies))
  File "/nextstrain/augur/augur/refine.py", line 72, in refine
    vary_rate=vary_rate, use_covariation=covariance, **kwarks)
  File "/usr/local/lib/python3.7/site-packages/treetime/treetime.py", line 178, in run
    self.make_time_tree(**tt_kwargs)
  File "/usr/local/lib/python3.7/site-packages/treetime/clock_tree.py", line 340, in make_time_tree
    self.init_date_constraints(clock_rate=clock_rate, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/treetime/clock_tree.py", line 293, in init_date_constraints
    one_mutation=self.one_mutation, branch_length_mode=self.branch_length_mode)
  File "/usr/local/lib/python3.7/site-packages/treetime/branch_len_interpolator.py", line 104, in __init__
    kind='linear', min_width=min_width)
  File "/usr/local/lib/python3.7/site-packages/treetime/distribution.py", line 155, in __init__
    self._fwhm = Distribution.calc_fwhm(self)
  File "/usr/local/lib/python3.7/site-packages/treetime/distribution.py", line 53, in calc_fwhm
    x_l, x_u = tmp[0], tmp[-1]
IndexError: index 0 is out of bounds for axis 0 with size 0

### Additional context
I am running this on the Terra Cloud computing system, or rather inside Terra.  The docker container I am using is called with this name: docker: "nextstrain/base" (if that helps)

**Comment by corneliusroemer** [comment]
Thanks @mdperry for this report! It could help debugging and fixing this if we had the input data you're running this with. Would it be possible to send it to cornelius.roemer@unibas.ch for example?

Can you confirm that the command runs fine without `--divergence-units 'mutations'` but errors when you add this parameter?

This could either be:
- invalid input data (in which case we should catch the error and provide a more informative message)
- augur bug
- treetime bug

@rneher @anna-parker do you have an idea what could be going on here?

**Comment by mdperry** [comment]
Yes, I can confirm that the command runs fine without the added option.  I should clarify that I am calling four augur commands in a row (will attach the code).  So for the job on Terra, when I say 'runs fine' I am not actually inspecting the output from augur refine, because it immediately gets fed into augur ancestral.  The JSON files produced in the end look very nice (to me).  Let's just say there were NO errors generated until I added the additional option/flag.

I have figured out from the log that the first 3 were successful and it was the fourth one processed that threw the exception so yes, I can Email you those files (there is nothing particularly sensitive or private about them).  I am also going to attach here for this sample what the 'Before' JSON subtree generated by UShER and then the 'After' JSON subtree that has been clocked by augur (in case that information helps at all)
[After_augur_CDPH-subtree-1981.json.gz](https://github.com/nextstrain/augur/files/9243429/After_augur_CDPH-subtree-1981.json.gz)
[bash_loop_to_clock_trees_in_terra.wdl.gz](https://github.com/nextstrain/augur/files/9243427/bash_loop_to_clock_trees_in_terra.wdl.gz)
[Before_from_UShER_CDPH-subtree-1981.json.gz](https://github.com/nextstrain/augur/files/9243450/Before_from_UShER_CDPH-subtree-1981.json.gz)

**Comment by corneliusroemer** [comment]
Excellent, thanks @mdperry! With all this info it should be possible to debug. @anna-parker is on it!

**Comment by anna-parker** [comment]
Hi @mdperry, sorry I was only able to have a look at your files until just now. It seems that the issue is not actually coming from the `--divergence-units mutations` flag and with or without it `treetime` will sometimes fail on your input. The issue is actually due to the branch length of the branch leading to the root in the `.nwk` tree you sent us being very long, but the rest of the tree not actually being that diverged. As ancestral sequences are unknown for the root node and this position also has the least information (as it has no ancestor) it is assigned a sequence by stochastic sampling. In this case the sequence that is stochastically generated at the start of treetime given the input of the tree and sequences is so far away from the other sequences of the branches that when we try to optimize the branch lengths in later rounds our optimizer is at times out of bounds and returns extremely large values which lead to the error you saw. This is partially due to the fact that `augur` is using treetime's `joint` method of branch length optimization and this is not intended to be used on long branches - since treetime version 0.9 we are able to use the `marginal` method of branch length optimization on larger trees due to speed ups - which should improve this slightly. (There is a PR currently about this so hopefully that will happen soon: https://github.com/nextstrain/augur/pull/1018). However, it also seems that the initial tree topology is very far away from the actual tree and as treetime uses this as a starting point this leads to issues. You can rerun this code a couple of times and treetime will sometimes manage to optimize the branch lengths of the tree, but I am also not sure if you should put that much weight into these results. Sorry to not give you any better news! I would suggest potentially running the samples through another tree builder and then using them as input for augur. But thank you for bringing this issue to up- we will make sure that we better catch these cases in future and give better error handling!

**Comment by mdperry** [comment]
Hi @anna-parker,

Thank you for your detailed explanation.  I am not familiar with all the different things to consider when trying to "clock" a tree, but fortunately other people on our Project team do have the relevant expertise, so I will discuss all of the points you raised with them.
Now that you mention it, I do indeed recall glimpsing a warning about long branch lengths in the STDERR from one of the augur calls, but I could not tell if it was important, nor did I know how to address it.  I find it interesting that the run for this sample's tree would fail sometimes even without the --divergence-units option.
Just off the top of my head, one solution would to be to identify problematic/challenging trees and simply skip them, but I suspect that the downstream users might eventually get frustrated if they could not find the clocked version of that tree.
Since I am using UShER to generate the big tree, and then matUtils to extract the small trees, I am wondering if there are some parameters or settings at either of those steps that I can modify or add to deal with long branch trees (somehow?).  But that is not your concern; I will consult with my local guru, Angie, and the other folks on the UShER dev team.  It is possible that they were already aware of situations similar to this.

Thanks again.

**Comment by anna-parker** [comment]
Hi @mdperry was actually still talking about this with @corneliusroemer and there is potentially another issue to look into. (Which might be of use when you use `augur` on future trees.)

`treetime` expects trees where branch lengths are proportional to the average number of nucleotide or protein substitutions per site. In the case of your input I believe the branch lengths are proportionate to the number of mutations in a sequence. 

I took the input tree you gave me and then divided each branch length by the wuhan reference sequence length (so by 29903) and then gave this as input to `treetime`. Using this input tree `treetime` runs without any errors (I ran 50 instances to rule out any stochastic issues). 

After running `treetime` `augur refine` should then, using the input sequences, convert the `mutations_per_site` into `mutations` (per sequence). https://github.com/nextstrain/augur/blob/e8392b320123b1422f86a8d61db1711681d9862f/augur/refine.py#L263 However, this doesn't seem to be happening... I don't want to use your trees for a bug report but I will see if I can reproduce this on another tree tomorrow as this does seem like a bug in `augur`.
